fsquota: only generate pod uuid is nil

pacoxu opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:

Fixes #112081

#### Special notes for your reviewer:

The bug was caused by https://github.com/kubernetes/kubernetes/pull/107302 by me when I fixed https://github.com/kubernetes/kubernetes/issues/107272.

I updated the `pkg/volume/emptydir/empty_dir.go` part to fix the logic and add some comments in `pkg/volume/util/fsquota/project.go`.

#### Does this PR introduce a user-facing change?

```release-note
fix a 1.24 regression in kubelet: fix a bug of stoping rendering configmap when enabling fsquota monitoring
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

pacoxu committed:
fsquota: only generate pod uuid is nil

pacoxu commented:
/priority important-soon
I need do some more test on this later.

pacoxu commented:
- Test failed [without this fix #112626](https://github.com/kubernetes/kubernetes/pull/112626): This is a test running with LocalStorageCapacityIsolationFSQuotaMonitoring to beta.  https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/112626/pull-kubernetes-node-crio-cgrpv2-e2e/1572531675106119680/
- Test passed in this PR #112625: set LocalStorageCapacityIsolationFSQuotaMonitoring to beta  https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/112625/pull-kubernetes-node-crio-cgrpv2-e2e/1572529867889577984/

This can prove that this can fix the regression.

pacoxu commented:
/kind regression

/cc @rphillips @dashpole 
/assign @derekwaynecarr @dchen1107 @mrunalp

dims commented:
This PR has the label work-in-progress, please revisit to see if you still need this, please close if not

pacoxu commented:
The bug was caused by https://github.com/kubernetes/kubernetes/pull/107302 by me when I fixed https://github.com/kubernetes/kubernetes/issues/107272.
/cc @jsafrane @dashpole 

I updated the `pkg/volume/emptydir/empty_dir.go` part to fix the logic and add some comments in `pkg/volume/util/fsquota/project.go`.

jsafrane commented on pkg/volume/util/fsquota/quota_linux.go:
```diff
@@ -304,7 +304,7 @@ func SupportsQuotas(m mount.Interface, path string) (bool, error) {
 // If the pod UID is identical to another one known, it may (but presently
 // doesn't) choose the same quota ID as other volumes in the pod.
-func AssignQuota(m mount.Interface, path string, poduid types.UID, bytes *resource.Quantity) error { //nolint:staticcheck // SA4009 poduid is overwritten by design, see comment below
+func AssignQuota(m mount.Interface, path string, _ types.UID, bytes *resource.Quantity) error {
```
Please remove `poduid types.UID` parameter, the function is called only on few places.

pacoxu replied:
Updated.

jsafrane commented on pkg/volume/emptydir/empty_dir.go:
```diff
@@ -301,12 +302,17 @@ func (ed *emptyDir) assignQuota(dir string, mounterSize *resource.Quantity) erro
+				}
+				return err
+			} else {
+				klog.V(4).Infof("Attempt to check quota on dir %s failed: %v", dir, err)
```
When `err == nil && projid == common.BadQuotaID`, this will log "`... failed: <nil>`", which is odd. In what cases can this happen and what does it mean to an user?

pacoxu replied:
This seems to be impossible.
- `BadQuotaID QuotaID = 0`
<img width="979" alt="image" src="https://user-images.githubusercontent.com/2010320/210794185-d72c4d82-0fbf-4b6a-ab0e-06e766464486.png">

```
if err == nil && projid != common.BadQuotaID {
```
can change to 
```
if err == nil {
```

Should I change it?

jsafrane replied:
Yes please, it would be more obvious.

pacoxu replied:
Updated.

pacoxu commented:
/cc @rphillips

pacoxu commented:
/triage accepted

rphillips commented on pkg/volume/emptydir/empty_dir.go:
```diff
@@ -301,12 +301,17 @@ func (ed *emptyDir) assignQuota(dir string, mounterSize *resource.Quantity) erro
-			if err != nil {
-				klog.V(3).Infof("Set quota on %s failed %s", dir, err.Error())
+			_, err := fsquota.GetQuotaOnDir(ed.mounter, dir)
+			if err == nil {
```
It feels like the logic in this function is backwards to usual golang practices. The functionality doesn't look wrong; just had to read it twice and has nesting that doesn't need to be here.

wdyt  (pretty sure I preserved the logic, but please double check):

```golang
_, err := fsquota.GetQuotaOnDir(ed.mounter, dir)
if err != nil {
  klog.V(4).Infof("Attempt to check quota on dir %s failed: %v", dir, err)
  // this is not a fatal error so return nil
  return nil
}
klog.V(4).Infof("emptydir trying to assign quota %v on %s", mounterSize, dir)
if err := fsquota.AssignQuota(ed.mounter, dir, ed.pod.UID, mounterSize); err != nil {
  klog.V(3).Infof("Set quota on %s failed %s", dir, err.Error())
  return err
}
return nil
```

pacoxu replied:
This can reduce one level of nesting. It looks better.

Updated.

pacoxu commented:
/sig node

jsafrane commented:
/approve
From sig-storage perspective

bart0sh commented:
/triage accepted
/priority important-soon

rphillips commented:
/lgtm

pacoxu commented:
I opened cherry-pick PRs for v1.24-v.26. Before that, I want to confirm if there are any CIs with this feature enabled are still failing. I need to add one if there is no such case.

alex-matei commented:
It doesn't seem this patch fixes the issue.

GetQuotaOnDir call in emptyDir.assignQuota function doesn't do anything. It always returns QuotaID 0 (BadQuotaID) and error is nil if the folder doesn't have quota enabled. If quota is enabled the error is still nil, so the call and if seem unnecessary.

For pods with configmaps the problem still persists:

```
Feb 02 11:38:16 we-test-master-0 kubelet[166883]: E0202 11:38:16.389758  166883 nestedpendingoperations.go:348] Operation for "{volumeName:kubernetes.io/configmap/2897430a-11ee-40cb-a35c-42698d4aba2c-settings-vol-config podName:2897430a-11ee-40cb-a35c-42698d4aba2c nodeName:}" failed. No retries permitted until 2023-02-02 11:40:18.389742605 +0000 UTC m=+1670.569255285 (durationBeforeRetry 2m2s). Error: MountVolume.SetUp failed for volume "settings-vol-config" (UniqueName: "kubernetes.io/configmap/2897430a-11ee-40cb-a35c-42698d4aba2c-settings-vol-config") pod "ama-logs-v9h8c" (UID: "2897430a-11ee-40cb-a35c-42698d4aba2c") : requesting quota on existing directory /mnt/kubelet/pods/2897430a-11ee-40cb-a35c-42698d4aba2c/volumes/kubernetes.io~configmap/settings-vol-config but different pod 8a83336a-fea5-4b8d-b9f9-9670bd8d5491 086871d7-c370-482b-bdb8-9db78173dab2
Feb 02 11:38:16 srvless-we-test-master-0 kubelet[166883]: I0202 11:38:16.389776  166883 event.go:294] "Event occurred" object="kube-system/ama-logs-v9h8c" fieldPath="" kind="Pod" apiVersion="v1" type="Warning" reason="FailedMount" message="MountVolume.SetUp failed for volume \"settings-vol-config\" : requesting quota on existing directory /mnt/kubelet/pods/2897430a-11ee-40cb-a35c-42698d4aba2c/volumes/kubernetes.io~configmap/settings-vol-config but different pod 8a83336a-fea5-4b8d-b9f9-9670bd8d5491 086871d7-c370-482b-bdb8-9db78173dab2"
```

The code in AssignQuota still generates a new random UID on each call and it compares with the previous one.

After kubelet restart the code still generates an error in createProjectID on every invocation:

```
Feb 02 11:51:20 we-test-master-0 kubelet[207133]: E0202 11:51:20.047407  207133 nestedpendingoperations.go:348] Operation for "{volumeName:kubernetes.io/configmap/2897430a-11ee-40cb-a35c-42698d4aba2c-osm-settings-vol-config podName:2897430a-11ee-40cb-a35c-42698d4aba2c nodeName:}" failed. No retries permitted until 2023-02-02 11:52:24.047399889 +0000 UTC m=+129.165734631 (durationBeforeRetry 1m4s). Error: MountVolume.SetUp failed for volume "osm-settings-vol-config" (UniqueName: "kubernetes.io/configmap/2897430a-11ee-40cb-a35c-42698d4aba2c-osm-settings-vol-config") pod "ama-logs-v9h8c" (UID: "2897430a-11ee-40cb-a35c-42698d4aba2c") : assign quota FAILED createProjectID /mnt/kubelet/pods/2897430a-11ee-40cb-a35c-42698d4aba2c/volumes/kubernetes.io~configmap/osm-settings-vol-config 0 failed attempt to reassign project ID for /mnt/kubelet/pods/2897430a-11ee-40cb-a35c-42698d4aba2c/volumes/kubernetes.io~configmap/osm-settings-vol-config
Feb 02 11:51:20 srvless-we-test-master-0 kubelet[207133]: I0202 11:51:20.047417  207133 event.go:294] "Event occurred" object="kube-system/ama-logs-v9h8c" fieldPath="" kind="Pod" apiVersion="v1" type="Warning" reason="FailedMount" message="MountVolume.SetUp failed for volume \"osm-settings-vol-config\" : assign quota FAILED createProjectID /mnt/kubelet/pods/2897430a-11ee-40cb-a35c-42698d4aba2c/volumes/kubernetes.io~configmap/osm-settings-vol-config 0 failed attempt to reassign project ID for /mnt/kubelet/pods/2897430a-11ee-40cb-a35c-42698d4aba2c/volumes/kubernetes.io~configmap/osm-settings-vol-config"
```

alex-matei commented on pkg/volume/emptydir/empty_dir.go:
```diff
@@ -301,12 +301,18 @@ func (ed *emptyDir) assignQuota(dir string, mounterSize *resource.Quantity) erro
 		} else if hasQuotas {
-			klog.V(4).Infof("emptydir trying to assign quota %v on %s", mounterSize, dir)
-			err := fsquota.AssignQuota(ed.mounter, dir, ed.pod.UID, mounterSize)
+			_, err := fsquota.GetQuotaOnDir(ed.mounter, dir)
```
This call and the check seems superfluous.

alex-matei commented on pkg/volume/util/fsquota/quota_linux.go:
```diff
@@ -319,11 +319,11 @@ func AssignQuota(m mount.Interface, path string, poduid types.UID, bytes *resour
-	poduid = types.UID(uuid.NewUUID()) //nolint:staticcheck // SA4009 poduid is overwritten by design, see comment above
-	if pod, ok := dirPodMap[path]; ok && pod != poduid {
-		return fmt.Errorf("requesting quota on existing directory %s but different pod %s %s", path, pod, poduid)
+	volumeuid := types.UID(uuid.NewUUID())
```
This seems only a change of variable name, how does it fix the problem?
